Nullable.md

```diff
@@ -0,0 +1,442 @@
+library's historical poor performance because of the compiler's non-optimized
+handling of union types. This Julep assumes that union types will be optimized
+such that poor performance is no longer a consequence of using a union type as
+the implementation of `?T`.
+
+Instead of concerns about inferior performance, one major argument against
+using `Union{T, Void}` is that a `Union{T, Void}` type can only be used
+reliably in code if the methods being called on instances of the union type
+are defined for both the `T` and `Void` sub-types of that union. Going forward,
+this document will refer to these two cases as the non-null and null cases,
+respectively.
+
+Because the `Union{T, Void}` representation uses a union type, it is possible
+for a method to be defined only over a subset of the values in the union. In
+particular, the possibility of directly calling `f(x::T)` on a value of type
+`Union{T, Void}` means that it is possible for code to handle the non-null case
+without being capable of handling the null case. We will refer to situations in
+which code handles only one of the non-null or null cases without handling the
+other case as failing to exhibit totality, a term used because the code does
+not define an execution path for the totality of all possible inputs.
+
+To ensure totality, users must either (a) use functions for which both
+`f(x::T)` and `f(x::Void)` are defined or (b) use higher-order functions
+that can take in a function with a method of the form `f(x::T)` and employ a
+default semantics for null values to handle the `f(x::Void)` case. In
+particular, we will refer to the pattern of extending an `f(x::T)` method to
+handle `f(x::Union{T, Void)` by assuming that `f(x::Void) = nothing` as
+lifting.
+
+The unique concern with `Union{T, Void}` is `f(x::T)` will often be defined
+because it is exists even in the parts of Julia that deal with only
+non-nullable types. When `T` is `Int`, this path is any function on integers.
+Many such functions exist and it is possible that users will assume that
+`f(x::Void)` will also exist because `f(x::Int)` exists.
+
+In DataArrays, the closely related problem of needing definitions of
+`f(x::NAtype) = NA` for all functions `f` was addressed by manually defining
+`f(x::NAtype) = NA` for all functions `f` that commonly occurred in practice.
+Although this piece-meal approach worked effectively in many cases because the
+majority of real code makes use of only a small number of core functions,
+there was always yet another function that to had be defined manually when a
+user found that `f(x::NAtype)` was not defined for their favorite function `f`.
+In this sense, the exact problems introduced by manually vectorizing scalar
+functions in Base Julia were repeated again when adding methods for `NAtype`:
+instead of describing a universal pattern using a generic abstraction, we
+repeated that pattern manually over and over again, while always discovering
+new cases that had been left out.
```

Comment:
This problem of manual redefinitions is not a criticism of the `Union{T, Void}` approach per se: it would rather justify hard-coding something like (pseudo-code) `(::F){F<:Function}(args...) = nothing` for all functions. (The next paragraph gives reasons why this would not be a good idea, but it's a different argument IMHO.)